documentation/puppet_server_metrics_performance.markdown

```diff
@@ -0,0 +1,78 @@
+
+There are two indicators in Puppet Server's metrics that can help you identify a request-handling capacity issue:
+
+-   **Average JRuby Wait Time:** This refers to the amount of time Puppet Server has to wait for an available JRuby to become available, and increases when each JRuby is held for a longer period of time, which reduces the overall number of free JRubies and forces new requests to wait longer for available resources.
+-   **Average JRuby Borrow Time:** This refers to the amount of time that Puppet Server "holds" a JRuby as a resource for a request, and increases because of other factors on the server.
+
+If wait time increases but borrow time stays the same, your Server infrastructure might be serving too many agents. This indicates that Server can easily handle requests but is receiving too many at once to keep up.
+
+If both wait and borrow times are increasing, something else on your server is causing requests to take longer to process. The longer borrow times suggest that Puppet Server is struggling more than before to process requests, which has a cascading effect on wait times. Correlate borrow time increases with other events whenever possible to isolate what activities might cause them, such as a Puppet code change.
+
+If you are setting up Puppet Server for the first time, start by increasing your Server infrastructure's capacity through additional JRubies (if your server has spare CPU and memory resources) or compile masters until your free JRubies are at least 0 and average number of free JRubies are at least 1. Once your system can handle its request volume, you can start looking into more specific performance improvements.
+
+#### Adding more JRubies
+
+If you need to add JRubies, remember that Puppet Server is tuned by default to use one fewer than your total number of CPUs, with a maximum of 4 CPUs, for the number of available JRubies. You can change this by setting `max-active-instances` in [`puppetserver.conf`][], under the `jruby-puppet` section. Don' set this value to something greater than the number of CPUs, because if there aren't enough processes available, it won't matter how many JRubies in the pool there are — the server won't be able to use them.
+
+Each JRuby also has a certain amount of memory overhead that it needs in order to be able to load both Puppet's Ruby code and your Puppet code. In other words, your available memory limits how much Puppet code you can process. As a general rule, adding a JRuby requires a bare minimum of 40MB of memory --- the amount of memory for the scripting container, plus Puppet's Ruby code, plus an additional 5MB) --- just to compile an almost-empty catalog. For real-world catalogs, add at least an additional 15MB per additional JRuby.
```

Comment:
A couple of notes about these memory numbers:

1. The numbers refer to persistent memory usage per JRuby.  The peak memory usage during operations like catalog compilations could grow to a much higher number.  I don't think we have any good measurements around peak memory usage to relay at this point, though.

2. For reference, the 40 MB number here was measured using JRuby 1.7 with Puppet Server and with the `jruby-puppet.compile-mode` setting set to `off`.  From early testing, we have seen memory usage be significantly higher with JRuby 9k and `jruby-puppet.compile-mode` set to `off` - roughly 61 MB or more - although that may improve in future JRuby 9k versions.  More memory can be consumed as well when the `jruby-puppet.compile-mode` is set to `jit` as well, with the tradeoff being potentially better performance.

3. I suppose the "real-world" measurement is pretty subjective.  In this case, we used the application of the `role::by_size::small` class from the [puppetlabs_perf_control](https://github.com/puppetlabs/puppetlabs-puppetserver_perf_control/blob/production/site/role/manifests/by_size/small.pp) repo to compute this value.  The catalog it produces installs a Tomcat and Postgres server along with 50 or so file resources related to managing user home directories.  Do you think it would be helpful to include any extra context about the catalog compilations we were measuring?